agent/attachment.go

```diff
@@ -0,0 +1,103 @@
+package agent
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/Sirupsen/logrus"
+	"github.com/docker/swarmkit/api"
+	"github.com/docker/swarmkit/identity"
+	"github.com/docker/swarmkit/log"
+	"golang.org/x/net/context"
+)
+
+// Listener allow to receive notifications about non executor attachment objects.
+type Listener interface {
+	// Notify notifies the listener about the updates over the executor attachment objects.
+	Notify([]*api.ExecutorAttachment)
+}
+
+// ExecutorAttachmentManager provides control over executor attachments on this node.
+type ExecutorAttachmentManager interface {
+	// CreateAttachment allows the node to request the allocation of resources
+	// needed for a container attachment on this node.
+	CreateAttachment(ctx context.Context, request *api.CreateExecutorAttachmentRequest) (*api.CreateExecutorAttachmentResponse, error)
+
+	// RemoveAttachment allows the node to request the release of
+	// the resources associated to the container attachment.
+	RemoveAttachment(context.Context, *api.RemoveExecutorAttachmentRequest) (*api.RemoveExecutorAttachmentResponse, error)
+
+	// Register allows clients to register to executor attachment notifications.
+	Register(Listener) (string, error)
+
+	// Leave leaves the notification clients pool.
+	Leave(string)
+}
+
+// ExecutorAttachmentManager returns the executor attachment management point.
+func (n *Node) ExecutorAttachmentManager() ExecutorAttachmentManager {
+	return n.agent
+}
+
+// Notifier provides notifications for non Task objects.
+type Notifier interface {
+	// Notify notifies the listeners about the updates over the executor attachment objects.
+	Notify(context.Context, []*api.ExecutorAttachment)
+}
+
+type notifier struct {
+	listeners map[string]Listener
+	sync.RWMutex
+}
+
+func newNotifier() *notifier {
+	return &notifier{
+		listeners: make(map[string]Listener),
+	}
+}
+
+func (n *notifier) Notify(ctx context.Context, eal []*api.ExecutorAttachment) {
+	for i, l := range n.listeners {
+		log.G(ctx).Debugf("Notifier: Notifying listener (%s):", i)
+		l.Notify(eal)
+	}
+}
+
+// Register allows clients to register for notifications.
+func (a *Agent) Register(l Listener) (string, error) {
```

Comment:
I don't understand why agent would implement this interface directly. It has almost nothing to do with the agent.